### ISSUE TITLE
Are switch expression patterns constant contexts?

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#51663]
The code:

```dart
  var a = switch (color) {
    Color.fromARGB(255, 187, 116, 111) => 'red',
    _ => 'unknown',
  };
```

Produces a diagnostic on `Color.fromARGB` (which is a `const` constructor):

> Color.fromARGB isn't a type.

If I add `const` in front of the constructor, then the error goes away.

This leaves me with a questions:

- Is this behaving correctly, is it invalid to not have `const`? If so, should the reporter error message be clearer?
- Or, since only constants are valid here, should `const` be implied? (this happens in some other contexts, although usually they already have `const` further up)

When I was stepping through this, `inConstantContext` seemed to be `true` for the expression.

@bwilkerson @lrhn

**Comment by lrhn** [comment]
Yes, it's behaving correctly. Constant constructor invocations now require a `const` in front.

That's because a construction invocation like `Foo(bar: 42, baz: "a")` is syntactically a valid *object pattern* matching a `Foo` and testing its `bar` and `baz` getters.
So the `const` is no longer implied, because it's used to distinguish between a constant and an object pattern.

This syntax does not match an object pattern, because the `255` isn't prefixed by a getter name, but all const constructor invocations require the `const`, so it's also syntactically not a constructor invocation.

The error message could be better. That's probably a matter of error recovery in the parser.

This is invalid as both a constructor invocation and an object pattern, so how do we say what's wrong with it grammatically?
(If we can wait and check what `Color.fromARGB` resolves to before giving the error message, it's probably a good idea.)

**Comment by bwilkerson** [comment]
IMO, if we can't wait for resolution, then the message should say that it is close to both but not either, and suggest how to make it one or the other.

@jensjoha For parser recovery

**Comment by DanTup** [comment]
@lrhn thanks for the explanation!

@bwilkerson do I understand correctly that line 2059 here is no longer correct?

https://github.com/dart-lang/sdk/blob/4733f9b754784b83406a5ba332d4ebc5974c1569/pkg/analyzer/lib/dart/ast/ast.dart#L2052-L2062

And the associated implementation case here should be removed?

https://github.com/dart-lang/sdk/blob/4733f9b754784b83406a5ba332d4ebc5974c1569/pkg/analyzer/lib/src/dart/ast/ast.dart#L4505-L4507

**Comment by bwilkerson** [comment]
> ... do I understand correctly that line 2059 here is no longer correct?

The comment probably needs to be updated, but it isn't as simple as removing line 2059. A switch case expression isn't a constant context in libraries with a language version of 3.x, but will still be a constant context in libraries with a language version of Dart 2.x. So the line needs to be updated with that information.

> And the associated implementation case here should be removed?

I believe that in Dart 3.0 we are now producing an instance of `SwitchPatternCase` rather than `SwitchCase`. I think that's true even for pre-3.0 code. If so the code needs to be updated to return a value based on the `patterns` feature. If I'm wrong, and we're still producing a `SwitchCase` for older code, then nothing needs to change.

@scheglov To confirm the AST nodes being produced.

**Comment by scheglov** [comment]
```dart
// @dart = 2.19
void f(Object? x) {
  switch (x) {
    case const Object():
      break;
  }
}
```
Is parsed with
```
    SwitchCase
      keyword: case
      expression: InstanceCreationExpression
```

And
```dart
void f(Object? x) {
  switch (x) {
    case const Object():
      break;
  }
}
```
is parsed with
```
    SwitchPatternCase
      keyword: case
      guardedPattern: GuardedPattern
        pattern: ConstantPattern
          const: const
          expression: InstanceCreationExpression
```

**Comment by bwilkerson** [comment]
In which case the implementation is correct; it's only the comment that needs to be updated.

**Comment by DanTup** [comment]
Sorry, I didn't step through the method properly and had assumed it was that case from just scanning the method. The line that actually causes me to get `inConstantContext=true` here is actually the `ConstantPatternImpl` case:

https://github.com/dart-lang/sdk/blob/4733f9b754784b83406a5ba332d4ebc5974c1569/pkg/analyzer/lib/src/dart/ast/ast.dart#L4484-L4485

**Comment by bwilkerson** [comment]
That case might be unnecessary, but my guess is that it's harmless. I'd have to look closer or try some tests to see what would break to know for sure.

**Comment by DanTup** [comment]
I was working on a fix for the color picker to insert `const` when you change a color like `Colors.red` (which is `const`) to `Color.fromARGB(x,y,z)` but that case above causes it to appear to already be a const context, and so we don't insert `const` (resulting in invalid code).

I can tidy up my change and open a review so you can see it with more context (and either the failing test, or this case removed and it passing).

**Comment by bwilkerson** [comment]
Or just update `ExpressionImplTest` to test `inConstantContext` directly.

**Comment by DanTup** [comment]
I'm not sure I follow - the issue is in new code (and tests) I'm writing (to decide whether `const` needs to be inserted when changing `Colors.red` to `Color.fromARGB(...)` using the colour picker). `ExpressionImplTest` already tests `inConstantContext`.

It may be easier to discuss with the code changes - I've opened https://dart-review.googlesource.com/c/sdk/+/287980.

**Comment by bwilkerson** [comment]
I thought that you were telling me that the problem was in `inConstantContext`, but I think I was mistaken.

**Comment by DanTup** [comment]
> I thought that you were telling me that the problem was in `inConstantContext`, but I think I was mistaken.

I was/am suggesting the problem was in `inConstantContext` - the change in https://dart-review.googlesource.com/c/sdk/+/287980 is changing `ExpressionImpl.inConstantContext` (to handle children of `ConstantPatternImpl` differently).

(Your other comments make sense, will fix the CL up tomorrow - thanks!)

**Comment by DanTup** [comment]
Nothing to do here, the question was answered and the one place that wasn't doing the right thing fixed as part of c58d445fa8c65ec1b2652fd3099e28fdd44b0da2.
